crates/core/src/transport/packet_data.rs

```diff
@@ -113,51 +128,137 @@ impl<const N: usize> PacketData<N> {
         Ok(Self {
             data: buffer,
             size: buffer_len,
+            data_type : PhantomData,
         })
     }
+}
+impl<const N: usize> PacketData<AssymetricEncrypted, N> {
+    pub(super) fn encrypt_with_pubkey(data: &[u8], remote_key: &TransportPublicKey) -> Self {
+        _check_valid_size::<N>();
+        let encrypted_data: Vec<u8> = remote_key.encrypt(data);
+        debug_assert!(encrypted_data.len() <= MAX_PACKET_SIZE);
+        let mut data = [0; N];
+        data.copy_from_slice(&encrypted_data[..]);
+        Self {
+            data,
+            size: encrypted_data.len(),
+            data_type : PhantomData,
+        }
+    }
+}
 
-    pub(super) fn is_intro_packet(&self, other: &PacketData<N>) -> bool {
-        if self.size != other.size {
+impl<const N: usize> PacketData<UnknownEncrypted, N> {
+    pub(super) fn is_intro_packet<DT : DataType>(&self, actual_intro_packet: &PacketData<AssymetricEncrypted, N>) -> bool {
+        if self.size != actual_intro_packet.size {
             return false;
         }
         let mut is_intro_packet = true;
         // TODO: how many bytes do we need to check to be sure that it's not the intro packet?
         // for now we randomly check 64 bytes (intro_packet is 1500 bytes long)
         for i in (0..64).map(|_| thread_rng().gen_range(0..self.size)) {
             // TODO: use a fast rng here?
-            if self.data[i] != other.data[i] {
+            if self.data[i] != actual_intro_packet.data[i] {
                 is_intro_packet = false;
                 break;
             }
         }
         is_intro_packet
     }
+    
+    pub(super) fn convert_to_sym_enc(self) -> PacketData<SymmetricEncrypted, N> {
+        PacketData {
+            data: self.data,
+            size: self.size,
+            data_type: PhantomData,
+        }
+    }
+    
+    pub(super) fn convert_to_asym_enc(self) -> PacketData<AssymetricEncrypted, N> {
+        PacketData {
+            data: self.data,
+            size: self.size,
+            data_type: PhantomData,
+        }
+    }
 }
 
-impl<const N: usize> From<[u8; N]> for PacketData<N> {
-    fn from(data: [u8; N]) -> Self {
-        _check_valid_size::<N>();
-        Self { data, size: N }
+impl<const N: usize> PacketData<PlainText, N> {
+    pub fn from_bytes_to_plain(data: [u8; N], size : usize) -> Self {
+        Self {
+            data,
+            size,
+            data_type: PhantomData,
+        }
+    }
+}
+
+impl<const N: usize> PacketData<AssymetricEncrypted, N> {
+    pub fn from_bytes_to_asym_enc(data: [u8; N], size : usize) -> Self {
```

Comment:
the only conversion from a byte array to a PacketData should be to the UnknownEncrypted subtype